feat(protocol-designer): wireup wasteChute commands

ncdiehl11 opened the pull request:
closes [RAUT-767](https://opentrons.atlassian.net/browse/RAUT-767)

# Overview

This PR updates the pipetting compound commands to check the dropTipLocation. If the location is included in additionalEquipment, wasteChuteCommandsUtil is called with the correct arguments needed. This is plugged into returnTip and generateRobotStateTimeline.

We also update the robot tip state when the commands get emitted by updating forDropTipLocation.

# Test Plan

- write test to verify that returned commands from replaceTip called at wasteChute return a moveToAddressableArea, dropTipInPlace, and pickUpTip
- check tip state to ensure replaceTip at waste chute updates used tips
- verify that we no-op moveToAddressableArea and dropTipInPlace if no tip attached

# Changelog

- 

# Review requests

@jerader per collaboration

# Risk assessment



[RAUT-767]: https://opentrons.atlassian.net/browse/RAUT-767?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jerader committed:
fix(protocol-designer): loadName in loadLabware is now true loadName (#13903)

addressing RESC-161

jerader committed:
fix(protocol-designer): moving labware onto an adapter on the deck (#13914)

closes RAUT-847

ahiuchingau committed:
we need to make sure positions are updated after fast home failure(#13934)

ncdiehl11 committed:
debug

jerader committed:
add check for tip attached in moveToAddressableArea

jerader committed:
fix ternary for no tip attached

koji committed:
feat(app): add robot is busy banner to device details (#13932)

* feat(app): add robot is busy banner to device details

sfoster1 committed:
feat(api): Partial tip configuration (#13936)

Adds partial-tip configurations for all multichannel pipettes (8-channel and 96-channel).

The idea here is that you can tell the system to use only a subset of a pipette's tips. This will really help with things like tip usage, and enable using high-throughput pipettes to interact with labware that would not support the geometry of multiple tips.


---------

Co-authored-by: Laura Cox <31892318+Laura-Danielle@users.noreply.github.com>
Co-authored-by: Carlos <fernandez.carlos40@gmail.com>
Co-authored-by: Laura Cox <laura@opentrons.com>

ncdiehl11 committed:
feat(protocol-designer): wire up drop drop tip in waste chute commands

If we select the waste chute as our drop tip location, we need to generate different commands from
dropping tips in other locations. For the waste chute, we generate moveToAddressableArea followed by
dropTipInPlace commands.

ncdiehl11 committed:
check for wasteChute

ahiuchingau committed:
listen to the right messages (#13937)

koji committed:
refactor(app): replace RobotWorkSpace with BaseDeck in ProtocolSetup Liquids  (#13918)

* refactor(app): replace RobotWorkSpace with BaseDeck in ProtocolSetup Liquids

mjhuff committed:
feat(app, app-shell): add download progress to robot update flows (#13915)

* feat(app): add download progress to robot update flows

If the robot update cache files are still downloading, no longer is an error thrown. Instead, the
update modal renders and a downloading stage of the update flow begins. Additionally, release files
are downloaded and returned to the browser layer separately from system files, so users may see
release notes as soon as they are downloaded.

ryanthecoder committed:
chore(api): fix push-ot3 for some versions of linux that can't use the multifile scp command (#13911)

fsinapi committed:
fix(api, shared-data, robot-server): stringify all error info in run log (#13942)

* fix(shared-data, api): stringify exception details dict to prevent runlog pydantic validation errors
* Add protection for run store to return None if pydantic validation of a  state summary fails
* Log run store errors when they happen

shlokamin committed:
refactor(app): only show ODD runs loading screen when server returns 503 (#13940)

ncdiehl11 committed:
fix(app): update video assets for attach and detach 96-channel pipette (#13929)

closes RQA-1863

Update video assets for attach/detach 96-channel pipette due to stroke issue in existing animations

andySigler committed:
fix(hardware-testing): Improvements to daily-setup process after more testing (#13899)

SyntaxColoring committed:
feat(robot-server): Add HTTP endpoints for getting and setting the robot's deck configuration (#13894)

Adds placeholder implementations of two new HTTP endpoints PUT /deck_configuration and GET /deck_configuration

jerader committed:
refactor(protocol-designer): remove unneeded loadLabware commands from 8_0_0 migration (#13945)

koji committed:
refactor(app): update deckmap and deck configurator styling (#13946)

* refactor(app): update deck configurator styling (#13946)

fsinapi committed:
fix(hardware): error details should always be strings (#13948)

ncdiehl11 committed:
update tip state after dropTipInPlace, add test for replaceTip in wasteChute

ncdiehl11 committed:
fix(app): ODD LPC copy and styling (#13925)

closes RQA-1052
closes RQA-1353
closes RQA-1358

Fix copy for preparing space conditional on step index and module presence. Correctly style
ExitConfirmation step. Fix button alignment in TipConfirmation.

ncdiehl11 committed:
lint

b-cooper committed:
fix(app, robot_server): allow non-standard tip racks in tip length calibration from dashboard (#13891)

Mirror our approach for loading an alternate tip rack into a legacy calibration session from
deck/pipette-offset to tip length calibration

Closes RAUT-809

b-cooper committed:
fix(app): fix last run protocol sorting on protocol dashboard (#13905)

We were incorrectly calculating the last run timestamp of a given protocol, this fixes the logic to
report the correct value

Closes RAUT-843

jerader committed:
feat(protocol-designer): modal updates and deck map design tweaks (#13900)

closes RAUT-807 RAUT-816 RAUT-817

caila-marashaj committed:
fix(shared-data): remove requirement for current entry to TipHandlingConfigurations (#13944)

ncdiehl11 committed:
feat(app): add command text for new commands

adds command text with requisite params for dropTipInPlace, dispenseInPlace, blowOutInPlace, and
moveToAddressableArea

ncdiehl11 committed:
initial tests

ncdiehl11 committed:
initial tests

ncdiehl11 committed:
add tests for all inPlace + moveToAddressableArea commands

b-cooper committed:
chore(protocol-designer, labware-library): remove unused full story integration (#13892)

b-cooper committed:
feat(app): allow multi file upload for labware and protocols (#13898)

Within the file browser that launches when importing protocols or labware into the Opentrons App,
support multi file select.

Closes RAUT-218

ncdiehl11 committed:
factor PipettingRunTimeCommand type into AtLocation and InPlace

b-cooper committed:
refactor(app, api-client): add utilities for parsing deck config via addressable areas (#13947)

Instead of explicit load fixture commands, extrapolate the simplest possible setup for the deck
given the included addressable areas referenced in a set of protocol commands

Closes RAUT-853

ncdiehl11 committed:
reset unintended files out of scope of this PR

ncdiehl11 committed:
reset pipetting types for now

jerader committed:
feat(protocol-designer): wire up unused equipment logic for waste chute (#13957)

closes RAUT-814

ahiuchingau committed:
fix origin & target for fast home should not include other axes (#13955)

smb2268 committed:
feat(app): final copy and animations for golden tip LPC probe screens (#13941)

fix RAUT-810, RAUT-811

smb2268 committed:
fix(app): module calibration should only use calibrated pipette (#13956)

fix RQA-1879

b-cooper committed:
refactor(app): remove unused analytics structures (#13893)

ncdiehl11 committed:
stash fixture files

mjhuff committed:
fix(app): fix step meter tearing by using no animation (#13960)

ODD optimizations do not prevent step meter tearing on the physical ODD when step progress
regresses. Playing no animation solves the problem until we optimize ODD animations.

mjhuff committed:
fix(app): Fix desktop app update modals (#13935)

* feat(app): add release notes button on app & robot update modals

* fix(app): fix closing app update modal during robot downgrade

When downgrading a robot, users are first prompted if they'd like to update their app. When clicking
close, the downgrade should proceed instead of crashing the app.

shlokamin committed:
refactor(app): remove console log on release notes component (#13951)

sfoster1 committed:
fix(shared-data): Fix the pipette schemas, tests (#13949)

The tests that were supposed to ajv the pipette definitions weren't set
up properly, and the tests that were supposed to test that the tests
were set up properly weren't set up properly, so we weren't testing the
schemas so the schemas were wrong. Fix the schema-test setup, the
test-setup-schema-test setup, and the schema.

jerader commented on protocol-designer/src/timelineMiddleware/generateRobotStateTimeline.ts:
```diff
@@ -2,6 +2,8 @@ import takeWhile from 'lodash/takeWhile'
 import * as StepGeneration from '@opentrons/step-generation'
 import { commandCreatorFromStepArgs } from '../file-data/selectors/commands'
 import type { StepArgsAndErrorsById } from '../steplist/types'
+import { wasteChuteCommandsUtil } from '@opentrons/step-generation'
```
this import isn't actually needed since line 2 is importing it. You can call it by typing `StepGeneration.wasteChuteCommandsUtil`

jerader commented on protocol-designer/src/timelineMiddleware/generateRobotStateTimeline.ts:
```diff
@@ -65,22 +67,55 @@ export const generateRobotStateTimeline = (
-                _prevRobotState
-              ),
-          ]
+          return isWasteChute
```
What you have totally works, but I think this ternary can be simplified further by pulling out the `CurriedCommandCreator` Like this:

```
const dropTipCommand  = isWasteChute
          ? StepGeneration.curryCommandCreator(StepGeneration.wasteChuteCommandsUtil, {
              type: 'dropTip',
              pipetteId: pipetteId,
              addressableAreaName,
            })
          : StepGeneration.curryCommandCreator(StepGeneration.dropTip, {
              pipette: pipetteId,
              dropTipLocation,
            })
```
            
Then:
```
    if (!willReuseTip) {
          return [
            ...acc,
            (_invariantContext, _prevRobotState) =>
              StepGeneration.reduceCommandCreators(
                [curriedCommandCreator, dropTipCommand],
                _invariantContext,
                _prevRobotState
              ),
          ]

        }
      }
```

ncdiehl11 replied:
agree, this is more concise

brenthagen committed:
refactor(shared-data,app): migrate to deck definition v4 (#13931)

changes default OT-2/Flex deck definitions to v4 in shared-data, app, components, and protocol
designer

closes RAUT-781

---------

Co-authored-by: Brian Cooper <brian@opentrons.com>

ncdiehl11 committed:
clean up dropTip command logic

ncdiehl11 committed:
feat(app): add inPlace and moveToAddressableArea command text (#13958)

closes RAUT-851

Generate different commands from dropping tips in other locations if we select the waste chute as our drop tip location. For the waste chute, we generate moveToAddressableArea followed by dropTipInPlace commands.

ncdiehl11 committed:
fix(app): robot settings and wizard required equipment styling (#13928)

closes RQA-1635
closes RQA-1639

Add divider above first and below last wizard required equipment element. Add padding between
calibration description and download cal logs button

shlokamin committed:
feat(shared-data): add 8 moveable trashes (#13959)

closes RSS-397

SyntaxColoring committed:
refactor(api): Delete unused opentrons.containers module #13973

shlokamin committed:
refactor(shared-data): update movable trash offset from cutout fixture (#13976)

jerader committed:
feat(step-generation): wire up moveLabware into waste chute (#13950)

closes RAUT-778

jerader committed:
feat(protocol-designer): deck v4 touchups and staging area support (#13965)

closes RAUT-787, RAUT-805, RAUT-806, RAUT-850

mjhuff committed:
fix(app): fix protocol slideout whitescreen  (#13977)

Closes RQA-1887, RQA-1888

* fix(app): fix all ProtocolSlideout deck thumbnails displaying the same deckmap

* fix(app): fix white-screen when selecting failed analysis protocol in protocol slideout

shlokamin committed:
ci(labware-library): add build timeout of 30 minutes (#13978)

ncdiehl11 committed:
add fixtures for 1000ul tiprack and 96-channel adapter

ncdiehl11 committed:
update replaceTip tests for 96-channel in waste chute

ncdiehl11 committed:
update fixture tests to reflect 1000ul tiprack and adapter

ncdiehl11 committed:
fix json formatting

jerader commented on step-generation/src/__tests__/replaceTip.test.ts:
```diff
@@ -132,6 +138,44 @@ describe('replaceTip', () => {
+          wasteChuteId: {
+            name: 'wasteChute',
+            id: wasteChuteId,
+            location: 'D3',
```
since we migrated over to v4 deck migration, lets update the location to reflect a true `cutoutId`:

```suggestion
            location: 'cutoutD3',
```

jerader commented on step-generation/src/__tests__/replaceTip.test.ts:
```diff
@@ -205,4 +249,49 @@ describe('replaceTip', () => {
+          wasteChuteId: {
+            name: 'wasteChute',
+            id: wasteChuteId,
+            location: 'D3',
```
same here

```suggestion
            location: 'cutoutD3',
```

jerader reviewed (commented):
thanks for adding some test coverage and test fixtures. I left a few comments to make sure our waste chute's location is a correct `cutoutId` to match the v4 deck.

There are a few tests failing on the PR in Step-generation that don't fail locally, do you mind rebasing `edge` so we are up to date with the v4 deck changes?

ncdiehl11 committed:
update wasteChute location in tests to be cutoutD3 to reflect deck def v4

ncdiehl11 commented:
moved to [here](https://github.com/Opentrons/opentrons/pull/13981) after rebase disaster
